### ISSUE TITLE
Background energy use

### ISSUE THREAD

**Issue by jrosindell** [OneZoom/OZtree#122]
We've found that sometimes the app / page uses significant energy when it's meant to be idle.

This is flagged as an issues as ideally we want to solve it but I've used an additional 'nice to have' label to indicate it's not the highest priority and not part of the MVP that needs to be delivered here.

**Comment by jrosindell** [comment]
The approach for this would be to check where set timeout or set interval are used alongside the Chrome debugging tool. API calls and basic background maintenance should happen only once every second when idle. It's hard to solve this problem without being able to repeat it but maybe it will be resolved automatically when large polytomy performance issues are also solved.

**Comment by hyanwong** [comment]
Safari often complains about the OneZoom page using too much energy, even when in the background, and I think this is the same cause - which makes it quite repeatable

**Comment by hyanwong** [comment]
Also see https://github.com/OneZoom/OZtree/issues/16

**Comment by lentinj** [comment]
The problem here isn't a great mystery---onezoom has a permanent ~60Hz requestAnimationFrame timer, which currently is rebuilding the shapes object tree from scratch regardless.

https://github.com/OneZoom/OZtree/commit/205aa1cbfa3d981e26425ac2323abae3617b7754 above eases this by only recalculating the shapes object tree if we actually need it, but as a result the browser can actually hit 60Hz for those rAF calls, so it's not a dramatic saving in idle CPU time.

``need_refresh`` and friends needs to be rethought so we start and stop the rAF timer as required.

**Comment by hyanwong** [comment]
Ah, very glad you know what's going on! It's a bit beyond me, TBH. I think this would be a very good thing to address, if we have sufficient "Jamie time" (how much work would this be?). What do you think, @jrosindell ?

**Comment by jrosindell** [comment]
Ah, welldone @lentinj for figuring this out.  One thing that's important to be aware of though is that detection of mouse over and thus of mouse click and actions within the caves is something that's integrated with the drawing routine - shapes might detect if the mouse pointer is over them as part of the drawing process.  Perhaps this is a bad paradigm to follow but it seemed elegant initially to choose in the same part of code that does the drawing wether or not that object should be interpreted as a 'button'.  The low effort solution to all this would be to restart the rAF loop on mouse move as well as on touch even if the view doesn't move itself, then have a condition to stop the rAF loop e.g. if nothing changed (including no mouse move etc.) in the last x seconds.  Another thing worth considering is whether the rAF should drop from 60Hz to 10Hz if the view isn't changing and we're just listening for mouse over events - I think the user would notice nothing if we did this but on the other hand it could be unnecessary complexity if in the end the user interaction loop is going to be entirely separated from the drawing loop.

**Comment by lentinj** [comment]
> detection of mouse over and thus of mouse click and actions within the caves

Yep, this is why my answer to @hyanwong was delayed, I don't think the actual changes will be that big, but picking apart the ramifications (and testing afterwards) will be the tricky bit.

But I think we can switch to a model where just have a call to request a new frame, which we do when...

* We change the tree_state viewport
* There's a mouse move event
* The node-developing timer ticks

We then let rAF run for ~1s after the last event, after which it stops. Somewhat like what is going on here:-

https://www.html5rocks.com/en/tutorials/speed/animations/#debouncing-mouse-events

After this we start picking apart the other timers, to see if they can also idle out (e.g. there's no need for the node-developing timer to fire if there's nothing more to get). This would remove our crutch of the screen being redrawn every 1s, but hopefully we don't need it.

This means we can't use the existing rendering loop for any kind of animation, but I don't think that's necessarily a problem? I've not seen any so far.

> Perhaps this is a bad paradigm to follow

It's another reason that this should be broken up into many separate canvas/SVG layers, and the interactive full leaves end up in a SVG layer. Then we can get normal browser machinations to do the hover-over for us, removing a lot of legwork.

**Comment by jrosindell** [comment]
Thanks Jamie, I agree.  I'm not sure how much of this will get implemented as part of OTOP optimisation but I guess that the timing loops will be fixed whereas the multi layered canvas will not? If this is so then please be mindful of the fact that in the end we'll want both of the optimisations to be done even if part 2 gets pushed down the road for the next project.

**Comment by lentinj** [comment]
Yes, and I'd posit that getting the timing loop isn't a nice to have, especially if the app store police notice.

**Comment by lentinj** [comment]
The above modifies the refresh loop as discussed. ``start_refresh_loop`` is replaced with ``trigger_refresh_loop``, and then calls to this are liberally applied wherever tree-modifying events happen.

I'm a bit unsatisfied with how many are required, but the alternative is instrumenting the tree_state and responding to changes in properties, which reading the source wasn't the desired way to do things.

This definitely needs testing, since there may be code paths that I've not added the requisite poking to re-enable the interface for, after which nothing will seem to happen unless you wiggle the mouse.

now with regular onezoom you only see the idle ticks of the image/node fetcher. These should be possible to remove too, since if there's nothing to get then there won't be anything more to get until the tree_state changes, but not as big-a concern.

OTOP you have substantial load of the background animation now, but there's not much I can do about that.

**Comment by jrosindell** [comment]
@lentinj this all looks good to me!   I'd not worry about the 1Hz ticks of the image / node fetcher.  Regarding the race condition, I wonder if you could add `trigger_refresh_loop` into the image fetcher callback so that when an image that the 3s timeout only starts once all images on screen have been loaded.  Alternatively I guess you could add `reset_global_button_action(); this.renderer.refresh(this.root);` to the 1Hz ticker so that image drawing refresh still happens in time with possible background updates to the data.

**Comment by hyanwong** [comment]
Unfortunately I think this will require user testing - I don't think my automatic test suite will be able to see if this works, although I can give it a try.

**Comment by hyanwong** [comment]
When zooming out, "fake" leaves (i.e. that are not yet expanded) are not developed any more: see screenshot below
![screen shot 2018-09-14 at 07 47 57](https://user-images.githubusercontent.com/4699014/45534344-aac7f580-b7f2-11e8-99be-b8fe146b9b20.png)

**Comment by hyanwong** [comment]
> Unfortunately I think this will require user testing - I don't think my automatic test suite will be able to see if this works, although I can give it a try.

And indeed the functional tests pass with the latest 3.3. @lentinj should we update 3.3 from master, just to keep things clean / in sync?

**Comment by hyanwong** [comment]
> Good idea, sorry I forgot to. Feel free to merge, or I'll do it later.

Done. I have no opinions on the source of the bare branches - it's the first time I've looked properly at 3.3. @jrosindell might want to test responsiveness etc more thoroughly, as he's more used to the expected behaviour.

Thanks @lentinj for tackling this energy use issue so speedily.

**Comment by jrosindell** [comment]
@lentinj thanks this all feels like a good step forward - and yes I expect the bare branches are related to #108 but regardless of that they are definitely not wanted behaviour and should be possible to fix without regressing the solutions put forward in #108.

**Comment by jrosindell** [comment]
Following up on this branch....

we still need to fix the bare branches mentioned above.

Also (and possibly related to this) during flight the handling of unloaded parts of the tree is not very nice - especially on OTOP - large chunks of the tree replaced with very large and unsightly fake leaves - on the ordinary OneZoom view there is a similar issue with large sections being replaced with circles (rather than leaf shapes)

**Comment by jrosindell** [comment]
Other than this Naziha did get the feeling that energy use is high still from her tests.  I think this might simply be related more to the performance issues on flight #91 but still I'd like to be sure that at least energy use is fully throttled back when idle before closing this issue

**Comment by lentinj** [comment]
> I'd like to be sure that at least energy use is fully throttled back when idle before closing this issue

There's not really anything else that can be done at this stage:

* We can't stop the OTOP background soaking up CPU without stopping the animation. We can change the animation but that's another discussion. These will stop if a tab isn't visible anyway.
* We can't remove re-rendering on mouseover until we break up the rendering into layers, and have an SVG layer for high-res leaves. This is a fairly epic task that doesn't live here.
* We could stop the node and image detail timers, but we don't really save anything.

..and the bare branches is solved within #108.
